### PR TITLE
Including the test suite from flask-k8s

### PR DESCRIPTION
**Pull request by weiiwang01**
Migrate the entire test suite including the integration tests and unit tests from https://github.com/canonical/flask-k8s-operator to here.

### CODE REVIEW

**Comment by jdkandersson** [comment]
you might have the record for largest PR...

File: examples/flask/charmcraft.yaml
Comment (jdkandersson) [inline examples/flask/charmcraft.yaml:1]: Should this be in git? It seems it is the expanded `charmcraft.yaml`?

File: examples/flask/default_rock/rockcraft.yaml
Comment (jdkandersson) [inline examples/flask/default_rock/rockcraft.yaml:1]: Similar question, this seems to be not using the extension?

File: examples/flask/requirements.txt
Comment (jdkandersson) [inline examples/flask/requirements.txt:4]: shouldn't this just depend on xiilib?

File: examples/flask/src/charm.py
Comment (jdkandersson) [inline examples/flask/src/charm.py:30]: hmm this might be a problem since we have to depend on `ops` so essentially we need to pin the major version of `ops`. Any thoughts @benhoyt? Also, @gregory-schiano

File: examples/flask/default_rock/rockcraft.yaml
Comment (weiiwang01) [reply examples/flask/default_rock/rockcraft.yaml:1]: We can use the rockcraft extension, but `operator-workflows` doesn't support that. I think it's okay to include that later.

File: examples/flask/requirements.txt
Comment (weiiwang01) [reply examples/flask/requirements.txt:4]: Unfortunately, charmcraft can't include files outside the charm directory in the build process, so we can't directly declare the dependency here. The current build process is to build the charm, then inject the local `xiilib` directory into the charm zip file.

File: examples/flask/src/charm.py
Comment (weiiwang01) [reply examples/flask/src/charm.py:30]: `ops` is a dependency of `xiilib`, it will be automatically installed.

File: examples/flask/charmcraft.yaml
Comment (weiiwang01) [reply examples/flask/charmcraft.yaml:1]: `operator-workflows` doesn't support that yet, I will submit a pull request later.

File: examples/flask/requirements.txt
Comment (jdkandersson) [reply examples/flask/requirements.txt:4]: Ah I see

File: examples/flask/src/charm.py
Comment (jdkandersson) [reply examples/flask/src/charm.py:30]: ah yeh you are right, so it should be ok

**Review by jdkandersson — DISMISSED** [review]

File: .github/ISSUE_TEMPLATE/bug_report.yml
Comment (gregory-schiano) [inline .github/ISSUE_TEMPLATE/bug_report.yml:1]: Could be for later, but we should probably adapt the bug report template as this repo is not directly a charm (while still used in a charm) but a library used in specific circumstances

File: .gitignore
Comment (gregory-schiano) [inline .gitignore:163]: Is it relevant ?

File: tests/integration/flask/conftest.py
Comment (gregory-schiano) [inline tests/integration/flask/conftest.py:127]: Shouldn't the config be in `charmcraft.yaml` ?

File: tests/conftest.py
Comment (gregory-schiano) [inline tests/conftest.py:12]: Shouldn't we have a more generic name, as soon as we do django or fastapi, that will require some renaming

File: tests/integration/flask/requirements.txt
Comment (gregory-schiano) [inline tests/integration/flask/requirements.txt]: Should we wait for this to be merged before merging this PR ?

Comment (weiiwang01) [reply tests/integration/flask/requirements.txt]: Good idea, pull request created: https://github.com/charmed-kubernetes/pytest-operator/pull/121

File: tests/integration/flask/conftest.py
Comment (weiiwang01) [reply tests/integration/flask/conftest.py:127]: As juju requires the `config.yaml` file inside the charm. charmcraft actually splits the `charmcraft.yaml` into `config.yaml` and `metadata.yaml` etc during the charm build time.

File: .gitignore
Comment (weiiwang01) [reply .gitignore:163]: Yes, as the integration tests still need to build rocks and charms, I think it might be safe to include them here.

File: tests/conftest.py
Comment (weiiwang01) [reply tests/conftest.py:12]: I think for Django and other frameworks they will have their own options like `--django-app-image`?

File: .gitignore
Comment (gregory-schiano) [reply .gitignore:163]: Fair enough 👍

File: tests/integration/flask/conftest.py
Comment (gregory-schiano) [reply tests/integration/flask/conftest.py:127]: Shouldn't this be supported by Harness ? We should probably report that to Ben

File: tests/conftest.py
Comment (gregory-schiano) [reply tests/conftest.py:12]: I'm not sure this will be relevant and scalable, but let's keep it this way and adapt once we work on Django if that's necessary

File: tests/integration/flask/requirements.txt
Comment (gregory-schiano) [reply tests/integration/flask/requirements.txt]: I asked for another round of review in this PR so this unblocks us here

File: tests/integration/flask/conftest.py
Comment (benhoyt) [reply tests/integration/flask/conftest.py:127]: It is, at least since 2.5.0 (which this repo requires). See https://github.com/canonical/operator/pull/977

Comment (weiiwang01) [reply tests/integration/flask/conftest.py:127]: Unfortunately, this is an integration test, so the harness doesn't have much use here.

**Review by jdkandersson — DISMISSED** [review]

**Review by jdkandersson — APPROVED** [review]

**Review by gregory-schiano — APPROVED** [review]

**Pull request merged**
